benchmarks/spiegelman_et_al_2016_gcubed/spiegelman_material.cc

```diff
@@ -0,0 +1,680 @@
+                  // strain rate (often simplified as epsilondot_ii)
+
+                  const double edot_ii = compute_second_invariant(edot, min_strain_rate[c]);
+
+                  // Find effective viscosities for each of the individual phases
+                  // Viscosities should have same number of entries as compositional fields
+
+                  // Regularized Drucker Prager viscosity (see Spiegelman et al, 2016)
+                  composition_viscosities[c] = compute_viscosity(edot_ii,pressure,c,prefactor[c],true,min_visc[c],max_visc[c]);
+
+                  Assert(dealii::numbers::is_finite(composition_viscosities[c]),ExcMessage ("Error: Viscosity is not finite."));
+
+                  if (derivatives != NULL)
+                    {
+                      if (use_analytical_derivative)
+                        {
+                          //analytic
+                          if (c == 0  && composition_viscosities[c] <= max_visc[c] && composition_viscosities[c] >= min_visc[c])
+                            {
+                              const double drucker_prager_viscosity = compute_viscosity(edot_ii,pressure,c,prefactor[c],false,min_visc[c],max_visc[c]);
+                              const double regulaization_adjustment = (ref_visc * ref_visc)
+                                                                      / (ref_visc * ref_visc + 2.0 * ref_visc * drucker_prager_viscosity
+                                                                         + drucker_prager_viscosity * drucker_prager_viscosity);
+
+                              composition_viscosities_derivatives[c] = -regulaization_adjustment *
+                                                                       (drucker_prager_viscosity / (edot_ii * edot_ii)) * edot;
+
+                              if (use_deviator_of_strain_rate == true)
+                                composition_viscosities_derivatives[c] = composition_viscosities_derivatives[c] * deviator_tensor<dim>();
+
+                              composition_dviscosities_dpressure[c] = regulaization_adjustment *
+                                                                      ((dim == 3)
+                                                                       ?
+                                                                       6 * sin_phi[c] / (sqrt_3 * (3.0 + sin_phi[c]) * 2.0 * sqrt_half * edot_ii)
+                                                                       :
+                                                                       sin_phi[c] / (2.0 * sqrt_half * edot_ii));
+                            }
+                          else
+                            {
+                              composition_viscosities_derivatives[c] = 0;
+                              composition_dviscosities_dpressure[c] = 0;
+                            }
+                        }
+                      else
+                        {
+                          // finite difference
```

Comment:
you checked that the two ways of computing things yield the same results, correct?

If so, let's just always use the analytic way of computing the derivatives because it's so much cheaper. In that case, just remove (i) this piece of code and (ii) the variable that decides which method to use.

Reply:
As discussed there are small difference between the two methods, but that should not affect the convergence. Since this is only a benchmark material, I think it is good to leave in both methods as a reference how they both can be implemented. I extended the documentation.